inference/__init__.py

```diff
@@ -0,0 +1,492 @@
+            if self.in_perm is not None:
+                _x = _x[self.in_perm]
+            spqr_cuda.spqr_mul(
+                self.m,
+                self.n,
+                self.bits,
+                self.beta1,
+                self.beta2,
+                self.buff0,
+                self.row_offsets,
+                self.col_vals,
+                self.nnz,
+                _x,
+                y[(i * self.m):((i + 1) * self.m)],
+                FeatureFlag.SPARSE_FUSED_FP32)
+
+        return y.reshape((1, inner_dim, self.m))
+
+
+# Compression
+
+def compress(W, bit_count):
+    W = list_flatten(W)
+    value_count = W.shape[0]
+
+    # Values per bucket
+    values_per_bucket = 64 // bit_count
+
+    total_buckets = (value_count + values_per_bucket - 1) // values_per_bucket
+    out = torch.zeros((total_buckets), dtype=torch.int64)
+
+    # Position within bucket
+    buff_id = 0
+
+    # Current bucket
+    bucket_id = 0
+
+    # Current bucket
+    bucket = np.uint64(0)
+
+    for i in range(value_count):
+        bucket |= np.uint64(W[i]) << np.uint64(buff_id * bit_count)
+        buff_id += 1
+
+        if buff_id == values_per_bucket:
+            out[bucket_id] = bucket
+            buff_id = 0
+            bucket = np.uint64(0)
+            bucket_id += 1
+
+    if buff_id > 0:
+        out[bucket_id] = bucket
+
+    return out
+
+
+def _spqr_dequantize(p: SPQRModule, nnz):
+    deq_w = torch.zeros(p.m, p.n).half().contiguous()
+    spqr_cuda.spqr_dequantize_compressed(
+        p.m,
+        p.n,
+        p.bits,
+        p.beta1,
+        p.beta2,
+        p.buff0,
+        p.row_offsets,
+        p.col_vals,
+        nnz,
+        deq_w)
+    return deq_w
+
+
+def spqr_dequantize_dense_compressed(p: SPQRModule):
+    return _spqr_dequantize(p, 0)
+
+
+def spqr_dequantize_compressed(p: SPQRModule):
+    return _spqr_dequantize(p, p.col_vals.shape[0])
+
+
+def spqr_dequantize_dense(p: SPQRUncompressed):
+    deq_w = torch.zeros(p.m, p.n).float().contiguous()
+    spqr_cuda.spqr_dequantize_host(
+        p.m,
+        p.n,
+        p.bits,
+        p.W,
+        p.beta1,
+        p.beta2,
+        p.W_s,
+        p.W_z,
+        p.W_s_s,
+        p.W_s_z,
+        p.W_z_s,
+        p.W_z_z,
+        p.values,
+        p.row_offsets,
+        p.col_ids,
+        0,  # This makes sure that we only quantize the dense part
+        deq_w)
+    return deq_w
+
+
+def spqr_dequantize(p: SPQRUncompressed):
+    deq_w = torch.zeros(p.m, p.n).float().contiguous()
+    spqr_cuda.spqr_dequantize_host(
+        p.m,
+        p.n,
+        p.bits,
+        p.W,
+        p.beta1,
+        p.beta2,
+        p.W_s,
+        p.W_z,
+        p.W_s_s.float(),
+        p.W_s_z.float(),
+        p.W_z_s.float(),
+        p.W_z_z.float(),
+        p.values.float(),
+        p.row_offsets,
+        p.col_ids,
+        p.nnz,
+        deq_w
+    )
+    return deq_w
+
+
+def spqr_mul_host(p: SPQRUncompressed, x, y_gt, y, dequantized_w):
```

Comment:
hmmm... why do you need this?

Reply:
This function used to call in a C++-side host function which had an implementation of  SPQR matvec. Therefore, it was used a reference while testing out the multiplication.  

There is no use for it anymore.

Resolved by deleting.